DentOS_Framework/DentOsTestbed/src/dent_os_testbed/test/test_suite/functional/policer/test_policer_ipv4_func.py

```diff
@@ -0,0 +1,132 @@
+    assert out[0][dent]["rc"] == 0, f"Failed creating bridge."
+
+    await IpLink.set(input_data=[{dent: [{"device": bridge, "operstate": "up"}]}])
+    assert out[0][dent]["rc"] == 0, f"Failed setting bridge to state UP."
+
+    # 2. Set link up on interfaces on all participant ports. Enslave all participant ports to the bridge.
+    out = await IpLink.set(input_data=[{dent: [{
+        "device": port,
+        "operstate": "up",
+        "master": bridge
+    } for port in ports]}])
+    assert out[0][dent]["rc"] == 0, f"Failed setting link to state UP."
+
+    # 3. Create an ingress qdisc on first port connected to Ixia port and add a rule with random selectors
+    out = await TcQdisc.add(input_data=[{dent: [{
+        "dev": ports[0],
+        "direction": "ingress"}]}])
+
+    if traffic_type == "random":
+        traffic_type = random.choice(["ipv4", "l2"])
+
+    rule_config = {
+        "action": "police",
+        "want_ip": True if traffic_type == "ipv4" else None,
+        "want_port": True if traffic_type == "ipv4" else None,
+        "skip_sw": True,
+        "want_proto": True if traffic_type == "ipv4" else None,
+        "want_mac": True,
+        "want_tcp": True if traffic_type == "ipv4" else None,
+        "police": {"rate": frame_rate, "burst": frame_rate + 1000, "conform-exceed": "drop"}}
+
+    tc_rule = tcutil_generate_rule_with_random_selectors(port_with_rule, **rule_config)
+
+    out = await TcFilter.add(input_data=[{dent: [tc_rule]}])
+    assert out[0][dent]["rc"] == 0, "Failed to create tc rule"
+
+    out = await TcFilter.show(input_data=[{dent: [{
+        "dev": port_with_rule,
+        "direction": "ingress",
+        "options": "-j"}]}], parse_output=True)
+    assert out[0][dent]["rc"] == 0, "Failed to get tc rule"
+
+    # 4 Prepare a matching stream for the randomly selected selectors and with random generated size of packet.
+    streams = tcutil_tc_rules_to_tgen_streams(
+        {port_with_rule: out[0][dent]["parsed_output"]}, frame_rate_pps=frame_rate, frame_rate_type="line_rate")
+
+    dev_groups = tgen_utils_dev_groups_from_config((
+        {"ixp": tg_ports[0], "ip": "1.1.1.1", "gw": "1.1.1.10", "plen": 24},
+        {"ixp": tg_ports[1], "ip": "1.1.1.2", "gw": "1.1.1.10", "plen": 24}
+    ))
+    await tgen_utils_traffic_generator_connect(tgen_dev, tg_ports, ports, dev_groups)
+
+    await tgen_utils_setup_streams(tgen_dev, None, streams)
+
+    # 5. Send Traffic
+    await tgen_utils_start_traffic(tgen_dev)
+    await asyncio.sleep(30)
+
+    # 6 Verify RX rate on RX ports is as expected (rate is limited by the police pass rule).
+    stats = await tgen_utils_get_traffic_stats(tgen_dev, "Port Statistics")
+    for row in stats.Rows:
+        if row["Port Name"] == tg_ports[0]:
+            continue
+        assert float(frame_rate - (frame_rate * tolerance)) <= float(
+            row["Rx. Rate (bps)"]) <= float(frame_rate + (frame_rate * tolerance))
+
+    await tgen_utils_stop_traffic(tgen_dev)
```

Comment:
Do we need traffic running during rate verification?

Reply:
I believe yes
I think that field `Rx. Rate (bps)` calculates during the traffic running. When traffic stops -> rate disappears